### PR TITLE
Validation apps for API

### PR DESCRIPTION
**Pull request by MateuszGrabuszynski**
Working on validation applications for v2 API. Very early stage at the moment.

### CODE REVIEW

File: tests/validation/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# Media Communications Mesh API validation
```
Comment (ko80) [inline tests/validation/README.md:1]:
```suggestion
# Media Communications Mesh SDK API validation
```
Please specify which SDK API where appropriate in the document.

File: tests/validation/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+This part of the repository is to be used solely for holding validation applications for Media Communications Mesh API and other files related to validating mentioned API.
+
+## Compilation process
+
+> **Note:** At the moment, it is not automated to expedite the development process.
+
+In order to compile the apps, first compile the whole Media Communications Mesh using `build.sh` script from the root of the repo, then use following commands:
```
Comment (ko80) [inline tests/validation/README.md:9]:
```suggestion
In order to compile the apps, first compile the whole Media Communications Mesh using `build.sh` script from the root of the repo, then use the following commands:
```

File: tests/validation/README.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+- `cc -o sender_app sender_app.c -lbsd $(pwd)/../../_build/lib/libmcm_dp.so`
+- `cc -o recver_app recver_app.c -lbsd $(pwd)/../../_build/lib/libmcm_dp.so`
+
+## Usage
+
+Remember to properly start the Media Proxy application before running the sample apps, in order to make the transmission possible.
```
Comment (ko80) [inline tests/validation/README.md:16]:
```suggestion
Remember to properly start the Media Proxy application before running the sample apps in order to make the transmission possible.
```

File: tests/validation/README.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+| `-i`   | `--rcv_port`      | `char[5]`          | Receiver port number   | `9001`                         | `9001`                     |
+| `-s`   | `--send_ip`       | `char[253]`        | Sender IP address      | `"127.0.0.1"`                  | `"192.168.96.2"`           |
+| `-p`   | `--send_port`     | `char[5]`          | Sender port number     | `9001`                         | `9002`                     |
+| `-o`   | `--protocol_type` | `char[32]`         | Type of protocol       | `"auto"`                       | `"memif"`                  |
+| `-t`   | `--payload_type`  | `char[32]`         | Type of payload        | `"st20"`                       | `"rdma"`/`"st22"`/`"st30"` |
+| `-k`   | `--socketpath`    | `char[108]`        | Socket path            | `"/run/mcm/mcm_rx_memif.sock"` | `"/other/path.sock"`       |
+| `-d`   | `--interfaceid`   | `uint32_t`         | ID of an interface     | `0`                            | `1`                        |
+| `-l`   | `--loop`          | `bool`             | Whether to loop or not | `0` (false)                    | `1` (true)                 |
+
+### Usage examples
```
Comment (ko80) [inline tests/validation/README.md:42]: Add the step to setup the environment, where you enable the VF.

File: tests/validation/README.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+- `sudo media_proxy -d 0000:c0:11.0 -t 9991`
+- `sudo MCM_MEDIA_PROXY_PORT=9991 ./recver -w 1920 -h 1080 -f 30 -x yuv422p10le" -s 192.168.96.2 -p 9992 -b output.yuv`
+
+## Validation helpers
+
+An aside part with help information for validators.
+
+### Media Proxy hugepages requirement
+
+To meet Media Proxy hugepages requirement, as super user, add a number >10 to each of the files:
```
Comment (ko80) [inline tests/validation/README.md:58]:
```suggestion
To meet Media Proxy hugepages requirement, as a super user, add a number >10 to each of the files:
```

Comment (ko80) [inline tests/validation/README.md:58]: Can you explain this `>10` requirement?

File: tests/validation/README.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+```text
+EAL: No free 2048 kB hugepages reported on node 0
+EAL: No free 2048 kB hugepages reported on node 1
+EAL: No free 1048576 kB hugepages reported on node 0
+EAL: No free 1048576 kB hugepages reported on node 1
+```
+
+### Media Proxy device vfio-pci mode
+
+If a following message is presented
```
Comment (ko80) [inline tests/validation/README.md:74]:
```suggestion
If the following message is presented
```

File: tests/validation/README.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+### Usage examples
+
+Sender:
+- `sudo media_proxy -d 0000:c0:11.1 -t 9992`
+- `sudo MCM_MEDIA_PROXY_PORT=9992 ./sender -w 1920 -h 1080 -f 30 -x yuv422p10le" -s 192.168.96.1 -r 192.168.96.2 -i 9991 -b input.yuv`
+
+Receiver:
+- `sudo media_proxy -d 0000:c0:11.0 -t 9991`
+- `sudo MCM_MEDIA_PROXY_PORT=9991 ./recver -w 1920 -h 1080 -f 30 -x yuv422p10le" -s 192.168.96.2 -p 9992 -b output.yuv`
+
```
Comment (ko80) [inline tests/validation/README.md:51]: After `Usage example` add a sample expected output.

File: tests/validation/README.md (hunk 5 above)
Comment (MateuszGrabuszynski) [reply tests/validation/README.md]: I'll be honest, I have no idea how many are needed for this to run properly. As far as I remember, two was not enough.

File: tests/validation/common.c [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    fprintf(fp, "-t, --type=payload_type\t\t"
+                "Payload type (default: %s)\n",
+        DEFAULT_PAYLOAD_TYPE);
+    fprintf(fp, "-p, --port=port_number\t\t"
+                "Receive data from Port (default: %s)\n",
+        DEFAULT_RECV_PORT);
+    fprintf(fp, "\n");
+
+
+    if (sender == 1){
```
Comment (ko80) [inline tests/validation/common.c:68]:
```suggestion
    if (sender == 1) {
```

Comment (ko80) [inline tests/validation/common.c:68]: Can you do the following instead?
```suggestion
    if (sender) {
```

File: tests/validation/recver_app.c [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        clock_gettime(CLOCK_REALTIME, &ts_recv);
+        if (first_frame) {
+            ts_begin = ts_recv;
+            first_frame = false;
+        }
+
+        if (dump_fp) {
+            fwrite(buf->data, buf->data_len, 1, dump_fp);
+        } else {
+            // Following code are mainly for test purpose, it requires the sender side to
```
Comment (ko80) [inline tests/validation/recver_app.c:301]: Avoid using C++ style comment in the C file. Stick to /* my comment here */`.

File: tests/validation/sender_app.c [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+        }
+
+        printf("TX frames: [%d], FPS: %0.2f [%0.2f]\n", frame_count, fps, vid_fps);
+        printf("Throughput: %.2lf MB/s, %.2lf Gb/s \n", throughput_MB,  throughput_MB * 8 / 1000);
+
+        frame_count++;
+
+        if (frame_count >= total_num) {
+            break;
+        }
```
Comment (ko80) [inline tests/validation/sender_app.c:373]:
```suggestion
        if (frame_count >= total_num)
            break;
```

File: tests/validation/README.md (hunk 5 above)
Comment (ko80) [reply tests/validation/README.md]: Please verify with @Mionsz, @tszumski, and @Sakoram.

Comment (Sakoram) [reply tests/validation/README.md]: Here [MTL docs](https://github.com/OpenVisualCloud/Media-Transport-Library/blob/main/doc/run.md#4-setup-hugepage) recommends 4 GB. The actual needed amount depends from size of video frame and number of streams. 

Your recommendation of 10 pages of 2 MB is irrelevant, only 10 pages of 1 GB will make any difference. You can as well recommend to allocate only 1 GB pages. 

Another topic is that some time ago I noticed a problem when the allocating both page types, the MTL configuration then took a very long time. If nothing has changed, I recommend to allocate one type of huge pages. for example 8 x 1GB or 4096 x 2MB

Comment (MateuszGrabuszynski) [reply tests/validation/README.md]: I reconsidered the number of each kind of hugepages based on the documentation provided and landed on hard recommending 8 GB. We usually have enough memory present on the systems to allow us to make 4 GB of each type. Additionally, a note was added with a link from Marek.

---

If you allocate only, say 2 MB hugepages, you will still be prompted with
```text
EAL: No free 1048576 kB hugepages reported on node 0
EAL: No free 1048576 kB hugepages reported on node 1
```
From my past experience, it makes the test fail. And adding the 1 GB hugepages to the mix fixes the issue.

File: tests/validation/recver_app.c [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+static volatile bool keepRunning = true;
+
+void intHandler(int dummy)
+{
+    keepRunning = 0;
+}
+
+int main(int argc, char** argv)
+{
+    char recv_addr[MESH_IP_ADDRESS_SIZE] = DEFAULT_RECV_IP;
```
[the remark this answers is not in the document]
Comment (MateuszGrabuszynski) [reply tests/validation/recver_app.c]: That's a great catch, thanks! :) I've resolved that.

File: tests/validation/common.c [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+#define DEFAULT_RECV_IP "127.0.0.1"
+#define DEFAULT_RECV_PORT "9001"
+#define DEFAULT_SEND_IP "127.0.0.1"
+#define DEFAULT_SEND_PORT "9001"
+#define DEFAULT_FRAME_WIDTH 1920
+#define DEFAULT_FRAME_HEIGHT 1080
+#define DEFAULT_FPS 30.0
+#define DEFAULT_PAYLOAD_TYPE "st20"
+#define DEFAULT_LOCAL_FILE "data-sdk.264"
```
[the remark this answers is not in the document]
Comment (MateuszGrabuszynski) [reply tests/validation/common.c]: Changed the name of the example to use `.yuv` extension.

File: tests/validation/README.md (hunk 4 above)
Comment (MateuszGrabuszynski) [reply tests/validation/README.md:95]: Added a short instruction for VF creation.

**Comment by MateuszGrabuszynski** [comment]
Huge thanks for the reviewers! :rocket: The issues were addressed. No more changes are planned at the moment, unless something new comes up.

**Review by ko80 — APPROVED** [review]

File: tests/validation/recver_val.c [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+#include <getopt.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <time.h>
+#include <unistd.h>
+#include "mesh_dp.h"
+#include "common.c"
```
Comment (Sakoram) [inline tests/validation/recver_val.c:17]: Why did you decided to include .c file instead of building it and linking?
Common defines are usually in .h file.

File: tests/validation/README.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+## Validation helpers
+
+An aside part with help information for validators.
+
+### Media Proxy hugepages requirement
+
+> **Note:** [Minimal Media Transport Library setup](https://github.com/OpenVisualCloud/Media-Transport-Library/blob/main/doc/run.md#4-setup-hugepage) specifies setting up 4 GB hugepages altogether. Below paragraph is just a recommendation, which should be used for testing purposes.
+
+To ensure stable testing of Media Proxy, as super user, add a number 4 to `/sys/kernel/mm/hugepages/hugepages-1048576kB/nr_hugepages` (4 * 1 GB hugepages) and 2048 to `/sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages` (2048 * 2 MB hugepages).
```
Comment (Sakoram) [inline tests/validation/README.md:226]: The user could not understand what do you mean by `add a number 4 to the file`. Maybe add a command `echo 4 > /sys/kernel/mm/hugepages/hugepages-1048576kB/nr_hugepages` or rephrase it to e.g.
- "Write the number  `4` to the file"
- "Save the number  `4`  into the file"

Comment (MateuszGrabuszynski) [reply tests/validation/README.md]: Changed to a command.

File: tests/validation/recver_val.c (hunk 13 above)
Comment (MateuszGrabuszynski) [reply tests/validation/recver_val.c]: Pure laziness. I had to remind myself how to do that properly. Hopefully, how I have done it is the right way... ;)

Plus, found a few minor bugs, which I fixed as well in the meantime, so it was definitely worth it.

**Review by Sakoram — APPROVED** [review]
Probably more code could be moved to common.c e.g. whole argument definition and parsing. I'd also recommend to use some script or more complex building system that will automatically build everything with single `make` or `build.sh`

but I don't want to look too hard for problems, as this is a validation application and is likely to be tweaked in future.
